# References

* **[Rube22]**

   Rube *et al.* (2022),
   *Prediction of protein-ligand binding affinity from sequencing data with interpretable machine learning*,
   [Nature Biotechnology](https://doi.org/10.1038/s41587-022-01307-0).

* **[Wolf18]**

   Virshup *et al.* (2021)
   *anndata: Annotated data*,
   [bioRxiv](https://doi.org/10.1101/2021.12.16.473007).

* **[Bartlett21]**
   Bartlett *et al.* (2021)
   *Fusion of single-cell transcriptome and DNA-binding data, for genomic network inference in cortical development*
   [BMC Bioinformatics](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-021-04201-9)

* **[Klein2023]**

   Klein *et al.* (2023)
   *Klein, D. et al. Mapping cells through time and space with moscot.*
   [bioRxiv](https://www.biorxiv.org/content/10.1101/2023.05.11.540374v2)

* **[Hochgerner2018]**

   Hochgerner *et al.* (2018)
   *Conserved properties of dentate gyrus neurogenesis across postnatal development revealed by single-cell RNA sequencing*
   [Nature Neuroscience](https://www.nature.com/articles/s41593-017-0056-2)

* **[Noack2022]**

   Noack *et al.* (2018)
   *Multimodal profiling of the transcriptional regulatory landscape of the developing mouse cortex identifies Neurog2 as a key epigenome remodeler*
   [Nature Neuroscience](https://www.nature.com/articles/s41593-021-01002-4)